### PR TITLE
self-assignment is a no-op

### LINKED ISSUE #556: self-assignment is a no-op
Basically all containers with `operator=` need to check for self-assignment and do nothing.

### PR DESCRIPTION
**Pull request by alandefreitas**
fix #556

### CODE REVIEW

File: include/boost/url/impl/params_encoded_ref.ipp [hunk 1]
```diff
@@ -40,7 +40,8 @@ params_encoded_ref::
 operator=(
     params_encoded_ref const& other)
 {
-    assign(other.begin(), other.end());
+    if (this != &other)
```
Comment (vinniefalco) [inline include/boost/url/impl/params_encoded_ref.ipp:43]: That's not enough you need to check `ref_.alias_of( other.ref_ )`

File: include/boost/url/static_url.hpp [hunk 2]
```diff
@@ -289,7 +289,8 @@ class static_url
     operator=(
         static_url const& u)
     {
-        copy(u);
+        if (this != &u)
```
Comment (vinniefalco) [inline include/boost/url/static_url.hpp:292]: This is good

File: test/unit/url_view.cpp [hunk 3]
```diff
@@ -56,6 +57,14 @@ class url_view_test
             BOOST_TEST_EQ(u1.data(), u1.data());
         }
 
+        // operator=(url_view_base const&)
+        {
+            url u1("x://y/z?#");
+            url_view u2 = u1;
+            u2 = u1;
+            BOOST_TEST_EQ(u1.data(), u1.data());
```
Comment (vinniefalco) [inline test/unit/url_view.cpp]: I thought I already did this test??

Comment (vinniefalco) [reply test/unit/url_view.cpp]: if you are going to have more than one test per member please group them together under one comment, like this:
```
        // operator=(url_view const&)
        {
            url_view u1("x://y/z?#");
            url_view u2;
            u2 = u1;
            BOOST_TEST_EQ(u2.data(), u1.data());
        }
        {
            // self-assignment
            url_view u1("x://y/z?#");
            u1 = u1;
            BOOST_TEST_EQ(u1.data(), u1.data());
        }
```

and add a comment explaining what the difference is

**Comment by alandefreitas** [comment]
merged

**Pull request closed without merging**